MobileCenterPush/MobileCenterPush/MSPush.m

```diff
@@ -0,0 +1,212 @@
+  [[self sharedInstance] didRegisterForRemoteNotificationsWith:deviceToken];
+}
+
++ (void)didFailToRegisterForRemoteNotificationsWith:(NSError *)error {
+  [[self sharedInstance] didFailToRegisterForRemoteNotificationsWith:error];
+}
+
+#pragma mark - MSServiceAbstract
+
+- (void)applyEnabledState:(BOOL)isEnabled {
+  [super applyEnabledState:isEnabled];
+  if (isEnabled) {
+    MSLogInfo([MSPush logTag], @"Push service has been enabled.");
+    if( !self.deviceTokenHasBeenSent ) {
+      [self registerForRemoteNotifications];
+    }
+  } else {
+    [[UIApplication sharedApplication] unregisterForRemoteNotifications];
+    MSLogInfo([MSPush logTag], @"Push service has been disabled.");
+  }
+}
+
+#pragma mark - Private methods
+
++ (void)resetSharedInstance {
+
+  // resets the once_token so dispatch_once will run again
+  onceToken = 0;
+  sharedInstance = nil;
+}
+
+- (void) registerForRemoteNotifications {
+  MSLogVerbose([MSPush logTag], @"Registering for push notifications");
+#if !(TARGET_OS_SIMULATOR)
+  if (floor(NSFoundationVersionNumber) <= NSFoundationVersionNumber_iOS_9_x_Max) {
+    UIUserNotificationType allNotificationTypes = (UIUserNotificationType) (UIUserNotificationTypeSound | UIUserNotificationTypeAlert | UIUserNotificationTypeBadge);
+    UIUserNotificationSettings *settings = [UIUserNotificationSettings settingsForTypes:allNotificationTypes categories:nil];
+    [[UIApplication sharedApplication] registerUserNotificationSettings:settings];
+  } else {
+#if defined(__IPHONE_10_0) && __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_10_0
+    UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+    UNAuthorizationOptions authOptions = (UNAuthorizationOptions) (UNAuthorizationOptionAlert | UNAuthorizationOptionSound | UNAuthorizationOptionBadge);
+    [center requestAuthorizationWithOptions:authOptions
+                          completionHandler:^(BOOL granted, NSError * _Nullable error) {}];
+    [center setDelegate:[[UIApplication sharedApplication] delegate]];
+#endif
+  }
+  [[UIApplication sharedApplication] registerForRemoteNotifications];
+#endif
+}
+
+#ifdef __IPHONE_8_0
+- (void)application:(UIApplication *)application didRegisterUserNotificationSettings:(UIUserNotificationSettings *)notificationSettings
+{
+  //register to receive notifications
+  [application registerForRemoteNotifications];
+}
+#endif
+
+- (NSString *)convertTokenToString:(NSData *)token {
+  if (!token)
+    return nil;
+  const unsigned char* dataBuffer = token.bytes;
+  NSMutableString *stringBuffer = [NSMutableString stringWithCapacity:(token.length * 2)];
+  for (NSUInteger i = 0; i < token.length; ++i) {
+    [stringBuffer appendFormat:@"%02x", dataBuffer[i]];
+  }
+  return [NSString stringWithString:stringBuffer];
+}
+
+- (void) sendDeviceToken: (NSString *)token {
+  MSPushLog *log = [MSPushLog new];
+  log.deviceToken = token;
+  [self.logManager processLog:log withPriority:MSPriorityHigh];
```

Comment:
Why is the priority high?
I know there are some places using MSPriorityHigh or MSPriorityDefault. The class already has `priority`, use self.priority.